### DISCUSSION TITLE
Why no mean_var_norm_emp.ckpt in checkpoint directory?

### DISCUSSION THREAD

**Discussion by Drawlone** [speechbrain/speechbrain#1646] [Q&A]
The pretrained ecapa_tdnn model offerd by speechbrain have mean_var_norm_emp.ckpt in checkpoint. But when I trained the whole model following readme, there is no this file. Does someone tell me why?

**Comment by Gastron** [comment]
Are you referring to this hyperparameter configuration [here](https://github.com/speechbrain/speechbrain/blob/develop/recipes/VoxCeleb/SpeakerRec/hparams/train_ecapa_tdnn.yaml)?

In that configuration, the InputNormalization checkpoint is saved as normalizer.ckpt - see this line: https://github.com/speechbrain/speechbrain/blob/b4efc906aca6d5ebf22a298bd0f4f44116b85fea/recipes/VoxCeleb/SpeakerRec/hparams/train_ecapa_tdnn.yaml#L175

**Comment by Drawlone** [comment]
@Gastron The pretrained model provided by speechbrain [here](https://drive.google.com/drive/folders/1-ahC1xeyPinAHp2oAohL-02smNWO41Cc?usp=sharing). It both includes mean_var_norm_emb.ckpt and normalizer.ckpt in 1986\save\CKPT+2021-02-27+12-48-32+00, and they are different ckpt files. 
But it doesn't generate mean_var_norm_emb.ckpt when I trained the model on my computer.  So I am confused about it.  
In addition, I need to make inference follow [this](https://huggingface.co/speechbrain/spkrec-ecapa-voxceleb). The mean_var_norm_emb.ckpt is necessary. Can I just replace normalizer.ckpt with mean_var_norm_emb.ckpt. I have no idea.........


> Are you referring to this hyperparameter configuration [here](https://github.com/speechbrain/speechbrain/blob/develop/recipes/VoxCeleb/SpeakerRec/hparams/train_ecapa_tdnn.yaml)?
> 
> In that configuration, the InputNormalization checkpoint is saved as normalizer.ckpt - see this line:
> 
> https://github.com/speechbrain/speechbrain/blob/b4efc906aca6d5ebf22a298bd0f4f44116b85fea/recipes/VoxCeleb/SpeakerRec/hparams/train_ecapa_tdnn.yaml#L175

**Comment by Gastron** [comment]
Ah. I see. This would be the normalization for the computed embeddings. Indeed I cannot find a place in the recipes where the normalization statistics would get computed. Perhaps @nauman-daw, @mravanelli or @hwidong-na can help?

This line defines the input normalization that gets used:
https://github.com/speechbrain/speechbrain/blob/b4efc906aca6d5ebf22a298bd0f4f44116b85fea/recipes/VoxCeleb/SpeakerRec/hparams/verification_ecapa.yaml#L72

It looks like the embedding normalization just gets run on the fly, without computing any statistics over the training data (which I assume is the normal way to do this)?

**Comment by mravanelli** [comment]
We do it in the speaker verification code. We save the one there.

**Comment by Gastron** [comment]
Ah yeah sorry. The normalization stats do get computed: https://github.com/speechbrain/speechbrain/blob/b4efc906aca6d5ebf22a298bd0f4f44116b85fea/recipes/VoxCeleb/SpeakerRec/speaker_verification_cosine.py#L266-L272

Is that right, though? I don't know about the verification conventions, but I would have thought that the normalization statistics should be computed over the training data, not the test data?

**Comment by Gastron** [comment]
Related issue: #1011

**Comment by zang5xixi** [comment]
how to fix this ???

**Comment by Gastron** [comment]
In speaker_verification_cosine.py you could save the normalization stats with something like this at the lines 266-272 (There's both runs for context in my snippet)
```python
... 
# First run
enrol_dict = compute_embedding_loop(enrol_dataloader) 
test_dict = compute_embedding_loop(test_dataloader) 

# Save normalization stats:
save_location = params["output_folder"] + "/mean_var_norm_emb.ckpt"
params["mean_var_norm_emb"]._save(save_location)

# Second run (normalization stats are more stable) 
enrol_dict = compute_embedding_loop(enrol_dataloader) 
test_dict = compute_embedding_loop(test_dataloader) 
...
```
